### PR TITLE
detect lightning htlc and unilateral close + fix multisig badge

### LINKED ISSUE #1091: multisig badge shows up even if not unlocked with multisig (special multisig configuration) + ideas about script detection / labeling P2SH and P2WSH
i made a 3 of 4 multisig that allows spending with a master key or a multisig configuration (`1 or (3 of 4)`)

when spending from this address with the master key the multisig badge (image) shows up

TXs:
unlocked with multisig: https://mempool.space/testnet/tx/a991d5485bece97b4f724508638e354e97cc97ea22228ca7bcd4d68c31bc6a1d
unlocked with master key: https://mempool.space/testnet/tx/f124530c94e94a234c4610940599189179a7bb05dc3e1bb62297e5bb436c17c6 (note that the `0` for the `OP_IF` is not visible due to https://github.com/mempool/mempool/issues/1089)

### PR DESCRIPTION
**Pull request by antonilol**
closes #1091

task list:

- [x] lightning contracts
- [x] fix multisig numbers
- [x] also detect uncompressed pubkeys

### CODE REVIEW

File: frontend/src/app/components/address-labels/address-labels.component.ts [hunk 1]
```diff
@@ -46,15 +46,21 @@ export class AddressLabelsComponent implements OnInit {
...
-      [
-        // {regexp: /^OP_DUP OP_HASH160/, label: 'HTLC'},
-        {regexp: /^OP_IF OP_PUSHBYTES_33 \w{33} OP_ELSE OP_PUSHBYTES_2 \w{2} OP_CSV OP_DROP/, label: 'Force Close'}
-      ].forEach((item) => {
-          if (item.regexp.test(this.vin.inner_witnessscript_asm)) {
-            this.lightning = item.label;
+      // https://github.com/lightning/bolts/blob/master/03-transactions.md#commitment-transaction-outputs
+      if (/^OP_IF OP_PUSHBYTES_33 \w{66} OP_ELSE OP_PUSHBYTES_(1 \w{2}|2 \w{4}) OP_CSV OP_DROP OP_PUSHBYTES_33 \w{66} OP_ENDIF OP_CHECKSIG$/.test(this.vin.inner_witnessscript_asm)) {
+          if (this.vin.witness[this.vin.witness.length - 2] == '01') {
+              this.lightning = 'Revoked Force Close';
```
Comment (antonilol) [inline frontend/src/app/components/address-labels/address-labels.component.ts:52]: force close/unilateral close

**Comment by softsimon** [comment]
Cool.
Could you list/link the a few example TXIDs that demonstrate/test the PR?

**Comment by antonilol** [comment]
This one got posted in the linked issue:

https://mempool.space/tx/bc0279a4b816b1ee2d6a2550c20e760ecb4f9da819fd6db732b064aba204bfa9

File: frontend/src/app/components/address-labels/address-labels.component.ts [hunk 2]
```diff
@@ -46,53 +46,62 @@ export class AddressLabelsComponent implements OnInit {
...
-      matches.push(match[index]);
+    const opN = ops.pop();
+    if (!opN.startsWith('OP_PUSHNUM_')) {
+      return;
     }
-    return matches;
+    const n = parseInt(opN.match(/[0-9]+/)[0]);
+    // pop n public keys
+    for (var i = 0; i < n; i++) {
+      if (ops.pop().length != 66) {
```
Comment (antonilol) [inline frontend/src/app/components/address-labels/address-labels.component.ts:87]: does not work for uncompressed keys, will be fixing this soon

**Comment by antonilol** [comment]
btw the problem why this didnt work before is that 33 bytes take up 66 hex characters, so `\w{66}` in regex

https://github.com/mempool/mempool/blob/181cb8e03f3cb1beeaf2ffe072eeb2e0c230a368/frontend/src/app/components/address-labels/address-labels.component.ts#L51

File: frontend/src/app/components/address-labels/address-labels.component.ts (hunk 2 above)
Comment (antonilol) [reply frontend/src/app/components/address-labels/address-labels.component.ts]: done

**Comment by antonilol** [comment]
final result
lightning:
[bc0279a4b816b1ee2d6a2550c20e760ecb4f9da819fd6db732b064aba204bfa9](https://mempool.space/tx/bc0279a4b816b1ee2d6a2550c20e760ecb4f9da819fd6db732b064aba204bfa9)
image
multisig detection doesn't have false positives anymore as it check every opcode

**Review by softsimon — APPROVED** [review]
tested ACK

**Pull request merged**